solc/CommandLineInterface.cpp

```diff
@@ -916,67 +918,134 @@ void CommandLineInterface::serveLSP()
 		solThrow(CommandLineExecutionError, "LSP terminated abnormally.");
 }
 
-void CommandLineInterface::link()
+static bool isBytecodeChar(char c)
 {
-	solAssert(m_options.input.mode == InputMode::Linker, "");
+	return isxdigit(c) || c == '_' || c == '$';
+}
 
+void CommandLineInterface::link(frontend::FileReader& fileReader, map<string, h160> const& libraries, bool& hasError, ostream& serr)
+{
 	// Map from how the libraries will be named inside the bytecode to their addresses.
 	map<string, h160> librariesReplacements;
 	int const placeholderSize = 40; // 20 bytes or 40 hex characters
-	for (auto const& library: m_options.linker.libraries)
+	for (auto const& library: libraries)
 	{
 		string const& name = library.first;
-		// Library placeholders are 40 hex digits (20 bytes) that start and end with '__'.
-		// This leaves 36 characters for the library identifier. The identifier used to
-		// be just the cropped or '_'-padded library name, but this changed to
-		// the cropped hex representation of the hash of the library name.
-		// We support both ways of linking here.
-		librariesReplacements["__" + evmasm::LinkerObject::libraryPlaceholder(name) + "__"] = library.second;
-
-		string replacement = "__";
-		for (size_t i = 0; i < placeholderSize - 4; ++i)
-			replacement.push_back(i < name.size() ? name[i] : '_');
-		replacement += "__";
-		librariesReplacements[replacement] = library.second;
+		librariesReplacements[name] = library.second;
 	}
 
-	FileReader::StringMap sourceCodes = m_fileReader.sourceUnits();
+	FileReader::StringMap sourceCodes = fileReader.sourceUnits();
+
 	for (auto& src: sourceCodes)
 	{
-		auto end = src.second.end();
-		for (auto it = src.second.begin(); it != end;)
+		// Map from how the libraries will be named inside the bytecode to their addresses.
+		map<size_t, string> linkReferences;
+		string bytecode;
+		vector<string> sourceLines;
+		boost::split(sourceLines, src.second, boost::is_any_of("\n"));
+
+		size_t bytecodeIndex;
+		bool bytecodeFound = false;
+		for (size_t i = 0; i < sourceLines.size(); ++i)
```

Comment:
Why not use the index variable as the loop counter?
```suggestion
		bool bytecodeFound = false;
		size_t lineIndex = 0;
		for (lineIndex; i < sourceLines.size(); ++lineIndex)
```

Alternatively, you could use range loop:
```suggestion
		bool bytecodeFound = false;
		size_t lineIndex = 0;
		for (string const& line: sourceLines)
```